File: XDrawer/Box.cs
﻿using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace XDrawer
{
    public class Box : TwoPointFigure
    {
        private bool _fillFlag;
        public Box(Popup popup, int x, int y)
            : base(popup, x, y)
        {
            _fillFlag = false;
        }
        public Box(Popup popup, int x1, int y1, int x2, int y2)
            : base(popup, x1, y1, x2, y2)
        {
            _fillFlag = false;
        }
        public override void setFill()
        {
            _fillFlag = !_fillFlag;
        }
        public override void draw(Graphics g, Pen pen)
        {
            if (_fillFlag == true)
            {
                Brush br = new SolidBrush(_color);
                g.FillRectangle(br, Math.Min(_x1, _x2), Math.Min(_y1, _y2), Math.Abs(_x2 - _x1), Math.Abs(_y2 - _y1));
                br.Dispose();
            }
            Color oldColor = pen.Color;
            pen.Color = _color;
            g.DrawRectangle(pen, Math.Min(_x1, _x2), Math.Min(_y1, _y2), Math.Abs(_x2 - _x1), Math.Abs(_y2 - _y1));
            pen.Color = oldColor;            
        }
    }
}

File: XDrawer/Box.cpp
#include "stdafx.h"
#include "Box.h"
#include "XDrawerView.h"

IMPLEMENT_SERIAL(Box, TwoPointFigure, 1)
Box::Box()
	:TwoPointFigure()
{	
	_fillFlag = FALSE;
}
Box::Box(int x1, int y1)
	:TwoPointFigure(x1, y1)
{	
	_fillFlag = FALSE;
}
Box::Box(int x1, int y1, int x2, int y2)
	:TwoPointFigure(x1, y1, x2, y2)
{
	_fillFlag = FALSE;
}
Box::~Box()
{
}
void Box::Serialize(CArchive& ar)
{
	TwoPointFigure::Serialize(ar);
}
void Box::draw(CDC* pDC)
{
	if (_fillFlag == FALSE) {
		CPen pen(PS_SOLID,1,_color);
		CPen *oldPen = pDC->SelectObject(&pen);
		pDC->Rectangle(m_x1, m_y1, m_x2, m_y2);
		pDC->SelectObject(oldPen);	
	} else {
		CPen pen(PS_SOLID,1,_color);
		CPen *oldPen = pDC->SelectObject(&pen);
		CBrush brush(_color);
		CBrush *oldBrush = pDC->SelectObject(&brush);
		pDC->Rectangle(m_x1, m_y1, m_x2, m_y2);
		pDC->SelectObject(oldBrush);
		pDC->SelectObject(oldPen);
	}	
}
void Box::setFill()
{
	_fillFlag = !_fillFlag;
}
Figure* Box::copy(CXDrawerView* view)
{
	Box *newBox = new Box(m_x1,m_y1,m_x2,m_y2);
	newBox->_fillFlag = _fillFlag;
	newBox->setColor(_color);
	newBox->setPopup(view->boxPopup);
	newBox->move(30,20);
	return newBox;
}
